### PR TITLE
web editor fix some HTML paste behaviors

### PR DESCRIPTION
**Pull request by sebgeelen**
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 1]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
-        const clipboardHtml = ev.clipboardData.getData('text/html');
+        const htmlClipboardData = ev.clipboardData.getData('text/html');
+        const plainTextData = ev.clipboardData.getData('text/plain');
+        let cleanHtmlData = null;
+        let shouldForceDeleteFoward = false;
+
+        if (htmlClipboardData) {
+            cleanHtmlData = this._cleanHtmlClipboardData(htmlClipboardData, cleanHtmlOptions);
+
+            const paragraphRegex = /^<p>(.+)<\/p>$/;
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2932]: Static regex, I realize this is splitting hairs but it's best to always keep those out of functions when possible so they're not recomputed every time.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 2]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
+            this.historyPauseSteps();
+            this.execCommand('insertHTML', cleanHtmlData, shouldForceDeleteFoward);
+            this.historyUnpauseSteps();
+            this.historyStep();
         } else {
-            const text = ev.clipboardData.getData('text/plain');
-            const splitAroundUrl = text.split(URL_REGEX);
+            const splitAroundUrl = plainTextData.split(URL_REGEX);
             const linkAttributes = this.options.defaultLinkAttributes || {};
+            const selectionIsInsideALink = !!closestElement(sel.anchorNode, 'a');
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2950]: This will only be true if the selection _starts_ in a link. What if it doesn't start in a link but ends in one?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
 
             for (let i = 0; i < splitAroundUrl.length; i++) {
+
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2953]: Unnecessary new line.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
                 // Even indexes will always be plain text, and odd indexes will always be URL.
-                if (i % 2) {
-                    const url = /^https?:\/\//gi.test(splitAroundUrl[i])
-                        ? splitAroundUrl[i]
-                        : 'https://' + splitAroundUrl[i];
-                    const youtubeUrl = YOUTUBE_URL_GET_VIDEO_ID.exec(url);
-                    const urlFileExtention = url.split('.').pop();
-                    const baseEmbedCommand = [
-                        {
+                // only allow images emebed inside an existing link. No other url or video embed.
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2962]: Capital.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 5]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
+                if (i % 2 && (isImageUrl || !selectionIsInsideALink)) {
+                    const baseEmbedCommand = [];
+                    if (!selectionIsInsideALink) {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2965]: This is great.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 6]
```diff
@@ -2930,28 +2982,23 @@ export class OdooEditor extends EventTarget {
...
+                    baseEmbedCommand.push({
+                        groupName: 'paste',
+                        title: 'Paste as text',
+                        description: 'Simple text paste.',
+                        fontawesome: 'fa-font',
+                        callback: () => {},
+                    });
 
+                    const undoAndApplyCommand = (callback) => {
+                        this.historyUndo();
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2996]: Uuuuh ok? What?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 7]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
+            const selectionIsInsideALink = !!closestElement(sel.anchorNode, 'a');
 
             for (let i = 0; i < splitAroundUrl.length; i++) {
+
+                const url = /^https?:\/\//gi.test(splitAroundUrl[i])
+                    ? splitAroundUrl[i]
+                    : 'https://' + splitAroundUrl[i];
+                const youtubeUrl = YOUTUBE_URL_GET_VIDEO_ID.exec(url);
+                const urlFileExtention = url.split('.').pop();
+                const isImageUrl = ['jpg', 'jpeg', 'png', 'gif'].includes(urlFileExtention);
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2959]: Maybe also svg? webp? ^^

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 8]
```diff
@@ -77,31 +77,38 @@ function insert(editor, data, isText = true) {
     }
     let nodeToInsert;
     const insertedNodes = [...fakeEl.childNodes];
+    let nodeToInsertId = 0;
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:80]: This is an index, not an Id.

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 9]
```diff
@@ -77,31 +77,38 @@ function insert(editor, data, isText = true) {
...
-                }
-                if (offset) {
-                    const [left, right] = splitElement(startNode.parentElement, offset);
-                    startNode = insertBefore ? right : left;
-                } else {
-                    startNode = startNode.parentElement;
+            if (nodeToInsertId === 0) {
+                makeContentsInline(nodeToInsert);
+                nodeToInsert = nodeToInsert.childNodes[0];
+            } else {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:86]: So all that changed here is this base case right? What does it cover? Could you add a comment?

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 10]
```diff
@@ -120,6 +127,13 @@ function insert(editor, data, isText = true) {
     newRange.setStart(lastPosition[0], lastPosition[1]);
     newRange.setEnd(lastPosition[0], lastPosition[1]);
     selection.addRange(newRange);
+
+    if (shouldForceDeleteFoward) {
+        // In some case when pasting a <p> inside another <p>,
+        // we need to merge them instead of creating a line break
+        // this force DeleteForward achieve this.
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:134]: > . This forced deleteForward achieves this.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 11]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
+            }
+        }
+
         if (files.length) {
             this.addImagesFiles(files);
-        } else if (clipboardHtml) {
-            this.execCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
+        } else if (cleanHtmlData && plainTextData !== cleanHtmlData) {
+            this.historyPauseSteps();
+            this.execCommand('insertHTML', cleanHtmlData, shouldForceDeleteFoward);
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2944]: Can't you do the delete forward here ?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 12]
```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
...
+            this.deleteRange(sel);
+            this.historyStep();
+        }
+
         const files = getImageFiles(ev.clipboardData);
-        const clipboardHtml = ev.clipboardData.getData('text/html');
+        const htmlClipboardData = ev.clipboardData.getData('text/html');
+        const plainTextData = ev.clipboardData.getData('text/plain');
+        let cleanHtmlData = null;
+        let shouldForceDeleteFoward = false;
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2927]: `shouldForceDeleteForward` (missing "r")

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 13]
```diff
@@ -443,8 +457,8 @@ function deleteTable(editor, table) {
 // absence of a strong test suite, so the whitelist stays for now.
 export const editorCommands = {
     // Insertion
-    insertHTML: (editor, data) => {
-        return insert(editor, data, false);
+    insertHTML: (editor, data, shouldForceDeleteFoward) => {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:460]: `shouldForceDeleteForward` (missing "r")

File: addons/web_editor/static/lib/odoo-editor/src/commands/commands.js [hunk 14]
```diff
@@ -45,7 +45,7 @@ import {
 const TEXT_CLASSES_REGEX = /\btext-[^\s]*\b/g;
 const BG_CLASSES_REGEX = /\bbg-[^\s]*\b/g;
 
-function insert(editor, data, isText = true) {
+function insert(editor, data, isText = true, shouldForceDeleteFoward = false) {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/commands.js:48]: `shouldForceDeleteForward` (missing "r")

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js [hunk 15]
```diff
@@ -163,8 +163,9 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false)
     }
 
     let node = this.childNodes[offset];
+    const nextSibblings = this.nextSibling;
     let firstBlockIndex = offset;
-    while (node && !isBlock(node)) {
+    while ( node && !isBlock(node) && (maxSizeNodeToMove === null || offset + maxSizeNodeToMove > firstBlockIndex)) {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:168]: Superfluous leading space.

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js [hunk 16]
```diff
@@ -163,8 +163,9 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false)
     }
 
     let node = this.childNodes[offset];
+    const nextSibblings = this.nextSibling;
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:166]: `nextSibling` (extra "b", extra "s")

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js [hunk 17]
```diff
@@ -65,7 +65,7 @@ Text.prototype.oDeleteBackward = function (offset, alreadyMoved = false) {
     setSelection(parentNode, firstSplitOffset);
 };
 
-HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false) {
+HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false, maxSizeNodeToMove = null) {
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:68]: That's a confusing new argument. Could you document it? Is that simply to avoid deleting blocks? If so, why?

Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:68]: That's a confusing new argument. Could you document it? Is that simply to avoid deleting blocks? If so, why?

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js [hunk 18]
```diff
@@ -183,7 +184,15 @@ HTMLElement.prototype.oDeleteBackward = function (offset, alreadyMoved = false)
...
         if (cType & CTGROUPS.BLOCK && (!alreadyMoved || cType === CTYPES.BLOCK_OUTSIDE)) {
-            cursorNode.oDeleteBackward(cursorOffset, alreadyMoved);
+            cursorNode.oDeleteBackward(cursorOffset, alreadyMoved, firstBlockIndex - offset);
+        } else if (!alreadyMoved) {
+            const cursorNodeNode = cursorNode.childNodes[cursorOffset];
+            const cursorNodeRightNode = cursorNodeNode ? cursorNodeNode.nextSibling : undefined;
+            if (cursorNodeRightNode &&
+                cursorNodeRightNode.nodeType === Node.TEXT_NODE &&
+                nextSibblings === cursorNodeRightNode) {
+                moveDest[0].insertBefore(document.createElement('br'), cursorNodeRightNode);
```
Comment (Zinston) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:194]: Comment please.
So if we're in the middle of text we add a br instead of deleting? I'm confused :-P

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 6 above)
Comment (Goaman) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2996]: `historyUndo` cannot work. Here is two examples: https://drive.google.com/file/d/1w8J4UNuEiF6edvsCoAWgZXNpWOcuyBw2/view

File: addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js [hunk 19]
```diff
@@ -0,0 +1,870 @@
+import {
+    BasicEditor, customErrorMessage,
+    testEditor
+} from "../utils.js";
+
+navigator.permissions.query({ name: "clipboard-write" }).then((result) => {
+    if (result.state === "granted" || result.state === "prompt") {
+        console.log("clipboard Write ok");
+    } else {
+        console.warn("clipboard Write Access bloqued !");
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js:10]: :ok_hand:

File: addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js [hunk 20]
```diff
@@ -0,0 +1,870 @@
+import {
+    BasicEditor, customErrorMessage,
+    testEditor
+} from "../utils.js";
+
+navigator.permissions.query({ name: "clipboard-write" }).then((result) => {
+    if (result.state === "granted" || result.state === "prompt") {
+        console.log("clipboard Write ok");
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js:8]: blocked

File: addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js [hunk 21]
```diff
@@ -12,6 +12,7 @@ function cycle(num, max) {
 
 export class Powerbox {
     constructor(options = {}) {
+        this._active = false;
```
Comment (Goaman) [inline addons/web_editor/static/lib/odoo-editor/src/powerbox/Powerbox.js:15]: what does it solve ?
shouldn't it be on another commit ?

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js (hunk 17 above)
Comment (Goaman) [inline addons/web_editor/static/lib/odoo-editor/src/commands/deleteBackward.js:68]: What are you fixing in oDeleteBackward ?

File: addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js [hunk 22]
```diff
@@ -0,0 +1,870 @@
+import {
+    BasicEditor, customErrorMessage,
+    testEditor
+} from "../utils.js";
+
+navigator.permissions.query({ name: "clipboard-write" }).then((result) => {
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/copyPaste.test.js:6]: :thinking: You probably right, It was used at the begening, but has been replaced by a mock event in `pasteData` ... 

I'll remove it.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 23]
```diff
@@ -47,8 +48,8 @@ import {
     YOUTUBE_URL_GET_VIDEO_ID,
     unwrapContents,
     peek,
-    rightPos,
-} from './utils/utils.js';
+    rightPos, parentsGet
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:51]: chill dude :stuck_out_tongue:

[the remark this answers is not in the document]
Comment (Zinston) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:51]: Oh I'm chill, wrong use of the goblin emoji that's all :D :dagger: :drop_of_blood:

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 24]
```diff
@@ -47,8 +48,8 @@ import {
...
+} from "./utils/utils.js";
```
[the remark this answers is not in the document]
Comment (Zinston) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:52]: :shrug:

**Comment by sebgeelen** [comment]
see #88817
